File: FinalProject/2_GraphAndTree/Baitoan7/TreeTraversal.cpp
#include <bits/stdc++.h>
using namespace std;

struct Node {
    string val;
    vector<Node*> children;
    Node(string v) : val(v) {}
};

unordered_map<string, Node*> nodes;

//(a) Preorder: Node -> Children
void preorder(Node* root) {
    if (!root) return;
    cout<<root->val<<" ";
    for (auto child : root->children) preorder(child);
}

//(b) Postorder: Children -> Node
void postorder(Node* root) {
    if (!root) return;
    for (auto child : root->children) postorder(child);
    cout<<root->val<<" ";
}

//(c) Top-down traversal (level order)
void topDown(Node* root) {
    if (!root) return;
    queue<Node*> q;
    q.push(root);
    while (!q.empty()) {
        Node* cur=q.front(); q.pop();
        cout<<cur->val<< " ";
        for (auto child : cur->children) q.push(child);
    }
}

//(d) Bottom-up traversal (reverse level order)
void bottomUp(Node* root) {
    if (!root) return;
    queue<Node*> q;
    vector<vector<string>> levels;
    q.push(root);
    while (!q.empty()) {
        int sz=q.size();
        vector<string> level;
        for (int i=0; i<sz; ++i) {
            Node* cur=q.front(); q.pop();
            level.push_back(cur->val);
            for (auto child : cur->children) q.push(child);
        }
        levels.push_back(level);
    }
    reverse(levels.begin(), levels.end());
    for (auto& level : levels)
        for (string& val : level) cout<<val<<" ";
}

int main() {
    int m;
    cin>>m;

    vector<pair<string, string>> edges;
    unordered_set<string> children_set;

    //nhập cạnh cha->con
    for (int i=0; i<m; ++i) {
        string u, v;
        cin>>u>>v;
        edges.emplace_back(u, v);
        children_set.insert(v);

        if (!nodes[u]) nodes[u]=new Node(u);
        if (!nodes[v]) nodes[v]=new Node(v);
        nodes[u]->children.push_back(nodes[v]);
    }

    //tìm root
    Node* root=nullptr;
    for (auto&[label, node] : nodes) {
        if (!children_set.count(label)) {
            root=node;
            break;
        }
    }

    if (!root) {
        cout<<"Khong tim thay root.";
        return 1;
    }

    cout<<"\n(a) Preorder  : ";
    preorder(root); cout<<"\n";

    cout<<"(b) Postorder : ";
    postorder(root); cout<<"\n";

    cout<<"(c) Top-down  : ";
    topDown(root); cout<<"\n";

    cout<<"(d) Bottom-up : ";
    bottomUp(root); cout<<"\n";

    return 0;
}
/*
5
A B
A C
A D
C E
C F

   A
 / | \
B  C  D
  / \
 E   F
*/

File: FinalProject/2_GraphAndTree/Baitoan7/TreeTraversal.py
from collections import deque

class Node:
    def __init__(self, val):
        self.val = val
        self.children = []

nodes = {}

m = int(input())
edges = []
children_set = set()

for _ in range(m):
    u, v = input().split()
    edges.append((u, v))
    children_set.add(v)

    if u not in nodes:
        nodes[u] = Node(u)
    if v not in nodes:
        nodes[v] = Node(v)
    nodes[u].children.append(nodes[v])

# Tìm root
root = None
for label in nodes:
    if label not in children_set:
        root = nodes[label]
        break

if root is None:
    print("Khong tim thay root.")
    exit(1)

def preorder(root):
    if not root:
        return
    print(root.val, end=' ')
    for child in root.children:
        preorder(child)

def postorder(root):
    if not root:
        return
    for child in root.children:
        postorder(child)
    print(root.val, end=' ')

def top_down(root):
    if not root:
        return
    q = deque([root])
    while q:
        cur = q.popleft()
        print(cur.val, end=' ')
        for child in cur.children:
            q.append(child)

def bottom_up(root):
    if not root:
        return
    q = deque([root])
    levels = []
    while q:
        sz = len(q)
        level = []
        for _ in range(sz):
            cur = q.popleft()
            level.append(cur.val)
            for child in cur.children:
                q.append(child)
        levels.append(level)
    for level in reversed(levels):
        for val in level:
            print(val, end=' ')

print("\n(a) Preorder  : ", end='')
preorder(root)
print()

print("(b) Postorder : ", end='')
postorder(root)
print()

print("(c) Top-down  : ", end='')
top_down(root)
print()

print("(d) Bottom-up : ", end='')
bottom_up(root)
print()